### PR TITLE
add simple function to count irreducible polynomials over 𝔽q

### PR DESCRIPTION
**Pull request by yyyyx4**
There is a well-known formula for counting irreducible polynomials of given degree $n$ over a finite field $\mathbb F_q$, using Möbius inversion. In this patch we add a simple function to evaluate this formula.

### CODE REVIEW

**Comment by fchapoton** [comment]
maybe rather put in src/sage/combinat/q_analogues.py ?

**Comment by yyyyx4** [comment]
Pardon my ignorance: Why?

**Comment by fchapoton** [comment]
because similar polynomials in q (cardinal of finite field) are already there..

and your method should take q as a parameter, and work with q an indeterminate

**Comment by maxale** [comment]
I have a few of comments:

1) why not use just one-liner:

```
return sum( moebius(n//d) * q**d for d in n.divisors() ) // n
```

instead of the current:

```
    r = ZZ.zero()
    for d in n.divisors():
        r += moebius(n//d) * q**d
    return r // n
```

2) The documentation should state that polynomials are monir.

3) There exists a formula for the number of monic irreducible _multivariate_ polynomials. It can be seen [here](https://dxdy-ru.translate.goog/post7034.html?_x_tr_sl=ru&_x_tr_tl=en&_x_tr_hl=en&_x_tr_pto=wapp#p7034) or in [this paper](http://arxiv.org/abs/0706.0157). I'd suggest to implement it and add the number of variables as yet another parameter with default value 1.

**Comment by yyyyx4** [comment]
Thanks for the suggestions: All done, except for the multivariate case, which is left for future work.

**Comment by GiacomoPope** [comment]
Depending on how important the multivariate case is, this could be modified to include `m = 1` as an optional parameter with a "NotImplementedError" with comments pointing to the references Max included. Personally I think this can be left for someone else to do though?

One thing I didn't understand was whether the formula in Bodin's work (which is only asymptotic) was exactly the same as Max's comment in the forum, or whether Max's comment and PARI implementation were precise?

Personally I think this univariate one is nice as it is, especially with the polynomial output when `q=None` and this recursive (potentially only asymptotic) formula for the multivariate one can be solved by someone else when they're interested in doing so?

**Comment by maxale** [comment]
> One thing I didn't understand was whether the formula in Bodin's work (which is only asymptotic) was exactly the same as Max's comment in the forum, or whether Max's comment and PARI implementation were precise?

@GiacomoPope: Bodin's result is not only asymptotic, Lemma 4 in his work gives an exact formula. That formula looks similar to mine (which was posted in the dxdy.ru forum). Historically, my formula appeared first (in 2006), but I considered it minor and did not bother to publish it elsewhere. PARI/GP code implements my formula and it was used to produce terms for many related sequences in the OEIS, some of which (such as [A006624](https://oeis.org/A115457)) I added back in 2006. I hope this explains the context.

Btw, A006624 links another relevant paper [Survey on counting special types of polynomials](http://arxiv.org/abs/1407.2970).

**Comment by GiacomoPope** [comment]
Thanks for the context Max. This helps. Potentially this function would be useful for people, but I don't think it needs to be included into this PR for this contribution to have value. It can be added at a later date without adding any additional work.

If you disagree though, then maybe porting your PARI code for this PR is something that needs to be done.

**Comment by maxale** [comment]
I'm not familiar enough with Sage development, and thus I'd appreciate it if someone can properly incorporate my code into this PR. Please find my extension to arbitrary positive `m` below. In the original code I've changed `q = ZZ['q'].gen()` to `q = QQ['q'].gen()` since in general for `m > 1` we may have an integer-valued polynomial but with rational coefficients. Also, I've changed `ZZ(n).divisors()` to `n.divisors()` because conversion `n = ZZ(n)` is already done as a first line. 

```
def number_of_irreducible_polynomials(n, q=None, m=1):
    n = ZZ(n)
    if n <= 0:
        raise ValueError('n must be positive')
    if q is None:
        q = QQ['q'].gen()       # for m > 1, we produce an integer-valued polynomial in q, but it does not necessarily have integer coefficients
    R = parent(q)
    if m==1:
        from sage.arith.misc import moebius
        r = sum((moebius(n//d) * q**d for d in n.divisors()), R.zero())
        if R is ZZ:
            return r // n
        return r / n
    elif m>1:                       # m > 1
        r = []
        for d in range(n):
            r.append( (q**binomial(d+m,m-1) - 1) // (q-1) * q**binomial(d+m,m) - sum(prod(binomial(r_+t-1,t) for r_,t in zip(r,p.to_exp(d))) for p in Partitions(d+1,max_part=d)) )
        return r[-1]
    else:
        raise ValueError('m must be positive')
```

**Comment by yyyyx4** [comment]
Thanks for writing this code @maxale; I've added it to the branch together with the corresponding documentation changes.

File: src/sage/combinat/q_analogues.py [hunk 1]
```diff
@@ -974,3 +974,77 @@ def q_stirling_number2(n, k, q=None):
...
+
+    In the multivariate case, formula from [Bodin2007]_,
+    independently [Alekseyev2006]_.
+    """
+    n = ZZ(n)
+    if n <= 0:
+        raise ValueError('n must be positive')
+    if q is None:
+        from sage.rings.rational_field import QQ
+        q = QQ['q'].gen()  # for m > 1, we produce an integer-valued polynomial in q, but it does not necessarily have integer coefficients
```
Comment (GiacomoPope) [inline src/sage/combinat/q_analogues.py]: For the example above, `1/12*q^12 - 1/12*q^6 - 1/12*q^4 + 1/12*q^2` has coefficients in QQ but the function is called with `m=1`, so I believe the statement should just be that evaluating this polynomial gives an integer but the coefficients can be in QQ?

File: src/sage/combinat/q_analogues.py [hunk 2]
```diff
@@ -974,3 +974,77 @@ def q_stirling_number2(n, k, q=None):
...
+    if m == 1:
+        from sage.arith.misc import moebius
+        r = sum((moebius(n//d) * q**d for d in n.divisors()), R.zero())
+        return r // n
+    elif m > 1:
+        from sage.functions.other import binomial
+        from sage.combinat.partition import Partitions
+        r = []
+        for d in range(n):
+            r.append( (q**binomial(d+m,m-1) - 1) // (q-1) * q**binomial(d+m,m) - sum(prod(binomial(r_+t-1,t) for r_,t in zip(r,p.to_exp(d))) for p in Partitions(d+1,max_part=d)) )
```
Comment (GiacomoPope) [inline src/sage/combinat/q_analogues.py]: More of a style thing, but one liners like this I find hard to read...


```py
def normalised_reducible(irreducible, d):
	"""
	Computes the number of normalised reducible polynomials
	of degree d given the a list of irreducible polynomials
	up to degree d-1
	"""
	res = 0
	for p in Partitions(d+1, max_part=d):
		tmp = 1
		for r, t in zip(irreducible, p.to_exp(d))
			tmp *= binomial(r+t-1,t)
		res += tmp
	return res
```

Then you could have this being slightly easier to follow:
```py
normalised = (q**binomial(d + m, m - 1) - 1) * q**binomial(d + m, m) // (q - 1)
normalised_reducible = normalised_reducible(r, d)
r.append(normalised - normalised_reducible)
```

But what is currently there is fine, even if a little opaque for someone reading this for the first time.

**Review by GiacomoPope — COMMENTED** [review]
Some comments, which can probably be taken or left.

File: src/sage/combinat/q_analogues.py (hunk 2 above)
Comment (grhkm21) [reply src/sage/combinat/q_analogues.py]: "normalised" should be defined. Also apparently Sage prefers American spelling c.f. #36581

Comment (maxale) [reply src/sage/combinat/q_analogues.py]: Maybe call the function `monic_irreducible`? I believe the term **monic polynomial** is more widely adopted than **normalized polynomial**.

Comment (grhkm21) [reply src/sage/combinat/q_analogues.py]: Yes definitely, I don't think "normalised" has a fixed meaning without context.

Comment (maxale) [reply src/sage/combinat/q_analogues.py]: > ```python
> 		tmp = 1
> 		for r, t in zip(irreducible, p.to_exp(d))
> 			tmp *= binomial(r+t-1,t)
> 		res += tmp
> ```

I do not mind unrolling some code for better readability, but the above piece looks an overkill to me. It implements the functionality of Python's `prod()` function and at the cost of creating an extra variable. This single-line looks much nicer and readable to me:

```
res += prod( binomial(r+t-1,t) for r, t in zip(irreducible, p.to_exp(d)) )
```

Comment (GiacomoPope) [reply src/sage/combinat/q_analogues.py]: This seems sensible to me

File: src/sage/combinat/q_analogues.py [hunk 3]
```diff
@@ -974,3 +974,95 @@ def q_stirling_number2(n, k, q=None):
...
+        Compute the number of monic reducible polynomials of degree `d`
+        given the numbers of irreducible polynomials up to degree `d-1`.
+        """
+        res = 0
+        for p in Partitions(d+1, max_part=d):
+            tmp = 1
+            for r, t in zip(irreducible, p.to_exp(d)):
+                tmp *= binomial(r+t-1, t)
+            res += tmp
+        return res
```
Comment (GiacomoPope) [inline src/sage/combinat/q_analogues.py]:
```suggestion
   def monic_reducible(irreducible, d):
        """
        Compute the number of monic reducible polynomials of degree `d`
        given the numbers of irreducible polynomials up to degree `d-1`.
        """
        res = 0
        for p in Partitions(d+1, max_part=d):
            res += prod( binomial(r+t-1,t) for r, t in zip(irreducible, p.to_exp(d)) )
        return res
```

**Comment by GiacomoPope** [comment]
@yyyyx4  I did something stupid in my suggestion while copy pasting and introduced whitespace issues!

File: src/sage/combinat/q_analogues.py [hunk 4]
```diff
@@ -974,3 +974,92 @@ def q_stirling_number2(n, k, q=None):
...
+
+   def monic_reducible(irreducible, d):
+        """
+        Compute the number of monic reducible polynomials of degree `d`
+        given the numbers of irreducible polynomials up to degree `d-1`.
+        """
+        res = 0
+        for p in Partitions(d+1, max_part=d):
+            res += prod( binomial(r+t-1,t) for r, t in zip(irreducible, p.to_exp(d)) )
+        return res
```
Comment (GiacomoPope) [inline src/sage/combinat/q_analogues.py]:
```suggestion
    def monic_reducible(irreducible, d):
        """
        Compute the number of monic reducible polynomials of degree `d`
        given the numbers of irreducible polynomials up to degree `d-1`.
        """
        res = 0
        for p in Partitions(d+1, max_part=d):
            res += prod( binomial(r+t-1,t) for r, t in zip(irreducible, p.to_exp(d)) )
        return res
```

**Review by mkoeppe — APPROVED** [review]
LGTM, appears to have addressed reviewer comments, passes tests.

**Pull request merged**